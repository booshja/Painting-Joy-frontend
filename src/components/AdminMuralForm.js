// dependencies
import React from "react";
import styled from "styled-components";
// components
import {
    StyledGreenSoloButton,
    StyledOutlineButton,
} from "../containers/styles/adminButtons";
// hooks
import { useForm } from "react-hook-form";
// breakpoints
import { breakpoints } from "../breakpoints";

const StyledForm = styled.form`
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    width: 90%;
    max-width: 700px;

    ${breakpoints("align-self", "", [{ 1024: "flex-start" }])}
`;

const StyledLabel = styled.label`
    font-family: "News Cycle", sans-serif;
    font-size: 1.4rem;
    letter-spacing: 1px;
    margin-bottom: 0.5rem;
`;

const StyledInput = styled.input`
    border: 2px solid #207070;
    border-radius: 4px;
    font-family: "News Cycle", sans-serif;
    font-size: 1.4rem;
    letter-spacing: 1px;
    margin-bottom: 0.5rem;
    width: 98%;
`;

const StyledField = styled.textarea`
    border: 2px solid #207070;
    border-radius: 4px;
    font-family: "News Cycle", sans-serif;
    font-size: 1.4rem;
    letter-spacing: 1px;
    margin-bottom: 1rem;
    resize: none;
    width: 98%;
`;

const StyledError = styled.p`
    color: #990000;
    font-family: "News Cycle", sans-serif;
    font-weight: 700;
    font-size: 1.2rem;
    letter-spacing: 1px;
    padding: 0.5rem;
`;

const StyledSoloBtn = styled(StyledGreenSoloButton)`
    border: 2px solid #207070;
`;

const StyledButtonContainer = styled.div`
    display: flex;
    justify-content: space-between;
    align-self: flex-end;
    width: 75%;
    align-items: center;

    ${breakpoints("width", "%", [{ 414: 60 }, { 768: 32 }, { 1024: 30 }])}
`;

const AdminMuralForm = ({
    handleDataSubmit,
    handleCancel,
    variant,
    preloadedValues,
}) => {
    // set up react-hook-form
    const {
        register,
        handleSubmit,
        formState: { errors },
    } = useForm({ defaultValues: preloadedValues });

    return (
        <StyledForm onSubmit={handleSubmit(handleDataSubmit)}>
            <StyledLabel htmlFor="title">Title:</StyledLabel>
            <StyledInput
                type="text"
                placeholder="Title"
                id="title"
                name="title"
                {...register("title", {
                    required: "Title is required",
                })}
            />
            {errors.title && <StyledError>{errors.title.message}</StyledError>}
            <StyledLabel htmlFor="description">Description:</StyledLabel>
            <StyledField
                type="text"
                placeholder="Description"
                id="description"
                name="description"
                {...register("description", {
                    required: "Description is required",
                })}
            />
            {errors.description && (
                <StyledError>{errors.description.message}</StyledError>
            )}
            <StyledButtonContainer>
                <StyledOutlineButton color="#DB9487" onClick={handleCancel}>
                    Cancel
                </StyledOutlineButton>
                <StyledSoloBtn type="submit">
                    {variant === "Add" && "Upload "}
                    {variant === "Edit" && " Edit "}
                    Images
                </StyledSoloBtn>
            </StyledButtonContainer>
        </StyledForm>
    );
};

export default AdminMuralForm;
