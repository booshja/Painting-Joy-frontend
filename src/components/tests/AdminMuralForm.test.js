import { render } from "@testing-library/react";
import AdminMuralForm from "../AdminMuralForm";

/** Smoke Test */
it("renders without crashing, variant add", () => {
    render(
        <AdminMuralForm
            handleDataSubmit={null}
            handleCancel={null}
            variant="Add"
            preloadedValues={null}
        />
    );
});

it("renders without crashing, variant edit, preloadedValues", () => {
    render(
        <AdminMuralForm
            handleDataSubmit={null}
            handleCancel={null}
            variant="Edit"
            preloadedValues={{
                title: "This is a title",
                description: "This is a description",
            }}
        />
    );
});

/** Snapshot Test */
it("matches snapshot, variant add", () => {
    const { asFragment } = render(
        <AdminMuralForm
            handleDataSubmit={null}
            handleCancel={null}
            variant="Add"
            preloadedValues={null}
        />
    );
    expect(asFragment()).toMatchSnapshot();
});

it("matches snapshot, variant edit, preloadedValues", () => {
    const { asFragment } = render(
        <AdminMuralForm
            handleDataSubmit={null}
            handleCancel={null}
            variant="Edit"
            preloadedValues={{
                title: "This is a title",
                description: "This is a description",
            }}
        />
    );
    expect(asFragment()).toMatchSnapshot();
});

/** RTL Tests */
it("renders correctly, variant add", () => {
    const { getByLabelText, getAllByRole } = render(
        <AdminMuralForm
            handleDataSubmit={null}
            handleCancel={null}
            variant="Add"
            preloadedValues={null}
        />
    );

    getByLabelText("Title:");
    getByLabelText("Description:");
    const buttons = getAllByRole("button");

    expect(buttons[0]).toHaveTextContent("Cancel");
    expect(buttons[1]).toHaveTextContent("Upload");
});

it("renders correctly, variant edit, preloadedValues", () => {
    const { getByDisplayValue, getAllByRole } = render(
        <AdminMuralForm
            handleDataSubmit={null}
            handleCancel={null}
            variant="Edit"
            preloadedValues={{
                title: "This is a title",
                description: "This is a description",
            }}
        />
    );

    getByDisplayValue("This is a title");
    getByDisplayValue("This is a description");
    const buttons = getAllByRole("button");

    expect(buttons[0]).toHaveTextContent("Cancel");
    expect(buttons[1]).toHaveTextContent("Edit");
});
