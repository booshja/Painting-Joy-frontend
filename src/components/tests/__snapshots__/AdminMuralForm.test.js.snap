// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot, variant add 1`] = `
<DocumentFragment>
  <form
    class="sc-dlnjwi iUMKXx"
  >
    <label
      class="sc-hKFxyN jXuugi"
      for="title"
    >
      Title:
    </label>
    <input
      class="sc-eCApnc gZPsrc"
      id="title"
      name="title"
      placeholder="Title"
      type="text"
    />
    <label
      class="sc-hKFxyN jXuugi"
      for="description"
    >
      Description:
    </label>
    <textarea
      class="sc-jSFjdj dCurNc"
      id="description"
      name="description"
      placeholder="Description"
      type="text"
    />
    <div
      class="sc-fujyAs kLrofT"
    >
      <button
        class="sc-gtsrHT cuBDVu"
        color="#DB9487"
      >
        Cancel
      </button>
      <button
        class="sc-bdnxRM sc-iCoGMd gcBejj cVXEZI"
        type="submit"
      >
        Upload Images
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`matches snapshot, variant edit, preloadedValues 1`] = `
<DocumentFragment>
  <form
    class="sc-dlnjwi iUMKXx"
  >
    <label
      class="sc-hKFxyN jXuugi"
      for="title"
    >
      Title:
    </label>
    <input
      class="sc-eCApnc gZPsrc"
      id="title"
      name="title"
      placeholder="Title"
      type="text"
    />
    <label
      class="sc-hKFxyN jXuugi"
      for="description"
    >
      Description:
    </label>
    <textarea
      class="sc-jSFjdj dCurNc"
      id="description"
      name="description"
      placeholder="Description"
      type="text"
    />
    <div
      class="sc-fujyAs kLrofT"
    >
      <button
        class="sc-gtsrHT cuBDVu"
        color="#DB9487"
      >
        Cancel
      </button>
      <button
        class="sc-bdnxRM sc-iCoGMd gcBejj cVXEZI"
        type="submit"
      >
         Edit Images
      </button>
    </div>
  </form>
</DocumentFragment>
`;
